### ISSUE TITLE
AnimationPlayer crash when playing in editor

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#24996]
**Godot version:**
3.1 beta1

**OS/device including version:**
Windows 10 x64

**Issue description:**
I animated a region_rect and when playing the animation in editor, Godot will crash. It happens only when pressing the play button, manually setting frame to end is ok and running the animation in game is ok too.

**Steps to reproduce:**
Edit reproduction project, open Obiekty/Drzwi.tscn, click AnimationPlayer and play the animation. The editor should crash at the end.

**Minimal reproduction project:**
[BugReport.zip](https://github.com/godotengine/godot/files/2757618/BugReport.zip)

EDIT:
Now I noticed that it's a tool script with animation end connected to queue free. I think that might be related.

**Comment by DualMatrix** [comment]
Ah so this actually doesn't seem to have anything to do with animating a region_rect. The actual cause seems to be that you connected `animation_finished` to a function in a `tool` script that calls `queue_free` which gets called even in the editor when the animation finished and apparently crashes the engine, to prevent this you can use `Engine.editor_hint` to check whether you're inside the editor.

Now, is the editor supposed to crash when one queue_frees a node from the scenetree in the editor, probably not right?

**Comment by reduz** [comment]
Yes, this will crash the editor. If you do this stuff with tool scripts you are on your own. 
Closing this issue.

**Comment by KoBeWi** [comment]
Eh, there should be a way to fix this. Like, catching the exception that crashes the editor and ignoring it when caused by tool script. Editor crashing is not a good thing, no matter what the user does to cause that.

**Comment by avencherus** [comment]
@KoBeWi Where it is possible I would agree, it does make it very difficult or infeasible to debug certain things when we don't get meaningful errors.

**Comment by KoBeWi** [comment]
Ok, I did some research and the issue happens only when you queue_free() root of currently opened scene. I then dived into source code and found out that changing e.g.
`if (get_scene_node()) {`
to
`if (get_scene_node() && ObjectDB::instance_validate(get_scene_node())) {`
will prevent editor from crashing on that particular line. However, you'd need to change many similar lines and that's rather unacceptable. The way to fix this is to make queue_free() set the root's pointer to null, so methods like get_scene_node() will return NULL, instead of invalid node. But I think I'll just make a new issue about this, because that's more broad than the bug brought here.
